File: R/stats-gsea.R
# GSEA layer: enrichment of curated metabolite-level pathway sets against a
# metabolite rank statistic. DESIGN.md §6.2, decisions.md #27/#35.

# Compute per-metabolite log2FC = mean(g2) - mean(g1) over log2-transformed
# concentrations. Returns a named numeric vector keyed by metabolite name.
# Outlier samples are excluded by `keep`.
.metabolite_log2fc <- function(prep, group) {
  keep <- !prep@sample_outliers
  if (sum(keep) < 2L) {
    stop("Fewer than two non-outlier samples available; cannot compute ",
         "metabolite log2FC.", call. = FALSE)
  }
  group <- group[keep]
  assay <- prep@sample@assay[keep, , drop = FALSE]
  metabs <- prep@sample@metabolite_names

  # log2 with adaptive pseudocount per metabolite (transpose so rows = metab).
  metab_log <- log2_with_pseudocount(t(assay))
  rownames(metab_log) <- metabs

  group <- droplevels(as.factor(group))
  if (length(levels(group)) != 2L) {
    stop("GSEA requires exactly two non-empty group levels; got ",
         length(levels(group)), ".", call. = FALSE)
  }
  g1 <- group == levels(group)[1L]
  g2 <- group == levels(group)[2L]

  m1 <- rowMeans(metab_log[, g1, drop = FALSE], na.rm = TRUE)
  m2 <- rowMeans(metab_log[, g2, drop = FALSE], na.rm = TRUE)
  fc <- m2 - m1

  # fgsea cannot handle NA / non-finite ranks; drop those metabolites.
  fc <- fc[is.finite(fc)]
  fc
}

#' GSEA over curated pathway sets
#'
#' Rank statistics from metabolite log2FC (group2 vs group1) are tested for
#' enrichment against the built-in curated metabolite-level pathway sets
#' (DESIGN.md §6.2). The rank statistic is computed on the
#' `PreprocessedData` concentrations; outlier samples are excluded.
#'
#' Direction information stored in `pathway_sets_meta` is *not* passed into
#' fgsea — it is metadata for downstream interpretation only
#' (decisions.md #40).
#'
#' @param prep A [PreprocessedData] object.
#' @param group Grouping variable as in [test_metabolites()].
#' @param domain One of the available pathway-set domains; see
#'   [list_pathway_sets()] (`"immunomet"`, `"reactome"`, `"wikipathways"`,
#'   `"smpdb"`, `"lion"`, `"source"`, `"health"`).
#' @param ... Passed to `fgsea::fgseaMultilevel()` (e.g., `minSize`,
#'   `maxSize`, `nPermSimple`).
#' @return The fgsea data.table result.
#' @export
gsea_pathway_sets <- function(prep, group, domain, ...) {
  .check_domain(domain)
  group <- .resolve_group(group, prep)
  pathways <- get_pathway_sets(domain)
  fc <- .metabolite_log2fc(prep, group)

  # Restrict each pathway to members present in the rank vector so fgsea
  # does not silently drop unknown names with confusing warnings.
  pathways <- lapply(pathways, function(m) intersect(m, names(fc)))
  pathways <- pathways[lengths(pathways) > 0L]
  if (length(pathways) == 0L) {
    stop("No pathway has any member overlapping the rank vector. ",
         "Check that `prep` covers the metabolites of domain '",
         domain, "'.", call. = FALSE)
  }

  fgsea::fgseaMultilevel(pathways = pathways, stats = fc, ...)
}

#' Convert an fgsea result to a flat data.frame for export
#'
#' [gsea_pathway_sets()] returns the raw fgsea `data.table`, whose
#' `leadingEdge` column is a list of character vectors. R's
#' `write.table()` / `write.csv()` cannot encode list columns and fail with
#' `unimplemented type 'list' in 'EncodeElement'`. This helper collapses
#' every list column with `paste(..., collapse = sep)` and returns a plain
#' data.frame so the result can be written to TSV/CSV directly. Non-list
#' columns are passed through unchanged. See DESIGN.md §6.2.3 and
#' decisions.md #53.
#'
#' @param res An fgsea-style result, typically from [gsea_pathway_sets()].
#' @param sep Single-character separator used to join list-column
#'   members. Default `";"`.
#' @return A `data.frame` with the same rows and columns as `res`; any
#'   list column (e.g. `leadingEdge`) becomes character.
#' @examples
#' \dontrun{
#'   res <- gsea_pathway_sets(prep, group = "Response", domain = "reactome")
#'   write.table(gsea_to_df(res), "reactome_gsea.tsv",
#'               sep = "\t", row.names = FALSE, quote = FALSE)
#' }
#' @export
gsea_to_df <- function(res, sep = ";") {
  out <- as.data.frame(res, stringsAsFactors = FALSE)
  is_list_col <- vapply(out, is.list, logical(1L))
  for (j in which(is_list_col)) {
    out[[j]] <- vapply(out[[j]],
                       function(x) paste(x, collapse = sep),
                       character(1L))
  }
  out
}

File: python/metabosetr/stats_gsea.py
"""Approach B: GSEA over curated pathway sets (port of ``stats-gsea.R``).

DESIGN.md §6.2. Rank statistic = per-metabolite log2FC (group2 vs group1)
over log2-transformed concentrations; outlier samples excluded. The R
version uses ``fgsea``; this port uses ``gseapy`` prerank (MIGRATION.md §2
decision B). Results are not numerically identical to fgsea, but are
deterministic given ``seed``.
"""

from __future__ import annotations

import numpy as np
import pandas as pd

from .pathways import _check_domain, get_pathway_sets
from .stats_feature import _resolve_group, _two_levels
from .transform import log2_with_pseudocount


def _metabolite_log2fc(prep, group) -> pd.Series:
    """Per-metabolite log2FC = mean(g2) - mean(g1); finite entries only."""
    keep = ~prep.sample_outliers
    if keep.sum() < 2:
        raise ValueError(
            "Fewer than two non-outlier samples available; cannot compute "
            "metabolite log2FC."
        )
    group = np.asarray(group, dtype=object)[keep]
    assay = prep.sample.assay[keep, :]
    metabs = list(prep.sample.metabolite_names)
    metab_log = log2_with_pseudocount(assay.T)  # rows = metabolites

    levels = _two_levels(group)
    g1 = group == levels[0]
    g2 = group == levels[1]
    with np.errstate(invalid="ignore"):
        m1 = np.nanmean(metab_log[:, g1], axis=1)
        m2 = np.nanmean(metab_log[:, g2], axis=1)
    fc = m2 - m1
    s = pd.Series(fc, index=metabs)
    return s[np.isfinite(s)]


def gsea_pathway_sets(prep, group, domain, seed: int = 0, **kwargs):
    """GSEA over the built-in curated pathway sets for ``domain``.

    ``seed`` pins the permutation RNG for reproducibility (decision B).
    Extra keyword args pass through to ``gseapy.prerank`` (e.g. ``min_size``,
    ``max_size``, ``permutation_num``). Returns the gseapy ``res2d``
    DataFrame.
    """
    import gseapy as gp

    _check_domain(domain)
    group = _resolve_group(group, prep)
    pathways = get_pathway_sets(domain)
    fc = _metabolite_log2fc(prep, group)

    present = set(fc.index)
    pathways = {sn: [m for m in members if m in present] for sn, members in pathways.items()}
    pathways = {sn: members for sn, members in pathways.items() if members}
    if not pathways:
        raise ValueError(
            "No pathway has any member overlapping the rank vector. Check that "
            f"`prep` covers the metabolites of domain '{domain}'."
        )

    rnk = fc.sort_values(ascending=False)
    params = dict(
        rnk=rnk,
        gene_sets=pathways,
        seed=seed,
        min_size=1,
        max_size=len(rnk),
        permutation_num=1000,
        outdir=None,
        no_plot=True,
        verbose=False,
    )
    params.update(kwargs)
    pre = gp.prerank(**params)
    return pre.res2d
